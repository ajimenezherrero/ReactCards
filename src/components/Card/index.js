import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import EditButton from '../Buttons/CircleButton'
import Wrapper from './Wrapper';
import WrapperDescription from './WrapperDescription';
import WrapperEditButton from './WrapperEditButton';
import WrapperTitle from './WrapperTitle';

class Card extends PureComponent {
  state = {
    active: false
  }

  handleClick = () => {
    const { onEdicClick, index } = this.props;

    onEdicClick({
      index
    });
  }

  handleMouseEnter = () => {
    this.setState({
      active: true
    });
  }

  handleMouseLeave = () => {
    this.setState({
      active: false
    });
  }

  renderEditButton() {
    const { active } = this.state;

    if (active) {
      return (
        <WrapperEditButton>
          <EditButton
            onClick={this.handleClick}
            backgroundColor="green"
            size="50px"
            fontSize="12px"
          >
            Edit
          </EditButton>
        </WrapperEditButton>
      );
    }
  }

  render() {
    const { description, image, title } = this.props;

    return (
      <Wrapper onMouseLeave={this.handleMouseLeave} onMouseEnter={this.handleMouseEnter}>
        <WrapperTitle image={image}>
          {title}
        </WrapperTitle>
        <WrapperDescription>
          {description}
        </WrapperDescription>
        {this.renderEditButton()}
      </Wrapper>
    );
  }
};

Card.defaultProps = {
  title: '',
  description: '',
  image: ''
};

Card.propTypes = {
  title: PropTypes.string,
  description: PropTypes.string,
  image: PropTypes.string
};

export default Card;
