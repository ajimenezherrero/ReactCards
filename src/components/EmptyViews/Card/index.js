import React from 'react';
import PropTypes from 'prop-types';

const CardsEmptyView = () => {
  return (
    <div>
      <label>EMPTY VIEW</label>
    </div>
  );
};

CardsEmptyView.propTypes = {
};

export default CardsEmptyView;
