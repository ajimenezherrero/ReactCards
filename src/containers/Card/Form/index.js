import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import Form from './Form';
import Wrapper from './Wrapper';

class CardsForm extends PureComponent {
  state = {
    title: '',
    description: '',
    image: ''
  }

  handleChange = event => {
    const changes = {};

    changes[event.target.name] = event.target.value;

    this.setState(changes);

  }

  handleSubmit = event => {
    const { onSubmit } = this.props;

    event.preventDefault();
    onSubmit(this.state);
  }

  render() {
    return (
      <Wrapper>
        <Form onSubmit={this.handleSubmit}>
          <label>
            Title:
            <input type="text" name="title" value={this.state.title} onChange={this.handleChange} />
          </label>
          <label>
            Description:
            <input type="text" name="description" value={this.state.description} onChange={this.handleChange} />
          </label>
          <label>
            Image(url):
            <input type="text" name="image" value={this.state.image} onChange={this.handleChange} />
          </label>
          <input type="submit" value="Submit" />
        </Form>
      </Wrapper>
    );
  }
};

CardsForm.propTypes = {};

export default CardsForm;
