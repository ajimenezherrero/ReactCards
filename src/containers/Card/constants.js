export const actionTypes = {
  ADD_CARD: 'CARD/ADD/CARD',
  SET_SHOW_ADD_CARD: 'CARD/SET/SHOW/ADD/CARD'
};
