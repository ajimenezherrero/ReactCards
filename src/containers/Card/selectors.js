import { createSelector } from 'reselect';

const getCardDomain = () => state =>  state.get('cardDomain');

const selectCards = () => createSelector(
  getCardDomain(),
  cardDomain => cardDomain.get('cards').toJS()
);

const selectShowAddCards = () => createSelector(
  getCardDomain(),
  cardDomain => cardDomain.get('showAddCards')
);

export default getCardDomain;
export {
  selectCards,
  selectShowAddCards
};
