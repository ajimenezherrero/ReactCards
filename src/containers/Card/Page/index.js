import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import Card from '../../../components/Card';

import Wrapper from './Wrapper';

class CardsPage extends PureComponent {
  handleOnEditClick = info => {
    console.log('INFO', info);
  }

  renderContent() {
    const { cards } = this.props;

    return cards.map(({ title, description, image }, index) => <Card key={index} index={index} title={title} description={description} image={image} onEdicClick={this.handleOnEditClick}/>);
  }

  render() {
    return (
      <Wrapper>
        {this.renderContent()}
      </Wrapper>
    );
  }
};

CardsPage.defaultProps = {
  cards: []
};

CardsPage.propTypes = {
  cards: PropTypes.array
};

export default CardsPage;
