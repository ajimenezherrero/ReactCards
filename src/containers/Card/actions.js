import { actionTypes as at } from './constants';

export const setShowAddCards = value => ({
  type: at.SET_SHOW_ADD_CARD,
  value
});

export const addCard = value => ({
  type: at.ADD_CARD,
  value
});
