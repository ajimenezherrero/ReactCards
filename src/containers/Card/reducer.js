import { combineReducers } from 'redux-immutable';
import { fromJS } from 'immutable';

import { actionTypes as at } from './constants';

export const initialState = fromJS({
  cards: [{
    title: 'title',
    description: 'description',
    image: 'https://s3-eu-west-1.amazonaws.com/cornerjob-cdn/acquisition/animatori.png'
  }],
  showAddCards: false
});

export const cardReducer = (state = initialState, { type, ...action}) => {
  switch (type) {
    case at.ADD_CARD:
      return state
        .set('cards', state.get('cards').push(action.value));
    case at.SET_SHOW_ADD_CARD:
      return state
        .set('showAddCards', action.value);
    default:
      return state;
  }
};

export default cardReducer;
