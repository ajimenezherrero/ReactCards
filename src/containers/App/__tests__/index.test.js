import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';

import { App, mapDispatchToProps } from '../index';
import { setShowAddCards, addCard } from '../../Card/actions'

jest.mock('../../Card/actions', () => ({
  setShowAddCards: jest.fn(),
  addCard: jest.fn()
}));

jest.mock('../../Card/Page', () => 'CardPage');
jest.mock('../../Card/Form', () => 'CardForm');
jest.mock('../../../components/Buttons/AddButton', () => 'AddButton');
jest.mock('../../../components/EmptyViews/Card', () => 'Card');
jest.mock('../../../components/Flexs/Flex', () => 'Flex');
jest.mock('../../../components/Modal', () => 'Modal');
jest.mock('../Wrapper', () => 'Wrapper');
jest.mock('../WrapperButton', () => 'WrapperButton');

const initialProps = {
  cards: []
}
const mockComponent = props => {
  const finalProps = { ...initialProps, ...props };

  return (
    <App { ...finalProps }/>
  );
};

describe('<App />', () => {
  it('should render component with empty cards', () => {
    const tree = create(mockComponent()).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with empty cards and show modal', () => {
    const props = {
      showAddCards: true
    }
    const tree = create(mockComponent(props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with cards', () => {
    const props = {
      cards: [{
        title: 'title',
        description: 'description',
        image: 'image'
      }]
    }
    const tree = create(mockComponent(props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('handleAddButtonClick', () => {
    it('should call onSetShowAddCards', () => {
      const onSetShowAddCards = jest.fn();
      const props = {
        onSetShowAddCards
      };
      const instance = shallow(mockComponent(props)).instance();

      instance.handleAddButtonClick();

      expect(onSetShowAddCards).toHaveBeenCalled();
    });
  });

  describe('handleSubmit', () => {
    it('should call onAddCard and onSetShowAddCards', () => {
      const onAddCard = jest.fn();
      const onSetShowAddCards = jest.fn();
      const props = {
        onAddCard,
        onSetShowAddCards
      };
      const instance = shallow(mockComponent(props)).instance();

      instance.handleSubmit();

      expect(onAddCard).toHaveBeenCalled();
      expect(onSetShowAddCards).toHaveBeenCalled();
    });
  });

  describe('mapDispatchToProps', () => {
    it('should dispatch setShowAddCards', () => {
      const { onSetShowAddCards } = mapDispatchToProps(jest.fn());

      onSetShowAddCards();

      expect(setShowAddCards).toHaveBeenCalled();
    });

    it('should dispatch addCard', () => {
      const { onAddCard } = mapDispatchToProps(jest.fn());

      onAddCard();

      expect(addCard).toHaveBeenCalled();
    });
  });
});
