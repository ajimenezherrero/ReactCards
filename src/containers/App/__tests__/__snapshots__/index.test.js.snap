// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render component with cards 1`] = `
<Wrapper>
  <CardPage
    cards={
      Array [
        Object {
          "description": "description",
          "image": "image",
          "title": "title",
        },
      ]
    }
  />
  <WrapperButton>
    <CircleButton
      onClick={[Function]}
    >
      Add
    </CircleButton>
  </WrapperButton>
</Wrapper>
`;

exports[`<App /> should render component with empty cards 1`] = `
<Wrapper>
  <Flex
    alignCenter={true}
    justifyCenter={true}
  >
    <Card />
  </Flex>
  <WrapperButton>
    <CircleButton
      onClick={[Function]}
    >
      Add
    </CircleButton>
  </WrapperButton>
</Wrapper>
`;

exports[`<App /> should render component with empty cards and show modal 1`] = `
<Wrapper>
  <Flex
    alignCenter={true}
    justifyCenter={true}
  >
    <Card />
  </Flex>
  <Modal>
    <CardForm
      onSubmit={[Function]}
    />
  </Modal>
  <WrapperButton>
    <CircleButton
      onClick={[Function]}
    >
      Add
    </CircleButton>
  </WrapperButton>
</Wrapper>
`;
